### PR TITLE
fix(airbyte-cdk): Add state workaround for legacy substreams

### PR DESCRIPTION
**Pull request by tolik0**
## What
This PR addresses an issue where the parent state is empty during the first sync when using `PerPartitionWithGlobalCursor`. It introduces state migration for substreams with incremental dependencies, ensuring that child state is copied to parent streams as needed.

## How
The set_initial_state method has been updated to check for incremental dependencies in parent streams. If parent_state is absent and at least one parent stream has an incremental dependency, the method will copy the child state to those parent streams.

## Can this PR be safely reverted and rolled back?

- [ ] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Review by maxi297 — COMMENTED** [review]
Can we add a test? The rest looks good to me

**Comment by tolik0** [comment]
[Regression test for Jira:
](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/manual/tolik0_airbyte-cdk_add-substream-state-migration/1731011617/9d0b96f5087f29d1b8d5712ba8de902f2526d4eb/source-jira/3.3.0-rc.1/output.html
)Screenshot from 2024-11-07 22-50-26

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py [hunk 1]
```diff
@@ -221,14 +223,46 @@ def set_initial_state(self, stream_state: StreamState) -> None:
...
+
+        # If `parent_state` doesn't exist and at least one parent stream has an incremental dependency,
+        # copy the child state to parent streams with incremental dependencies.
+        incremental_dependency = any([parent_config.incremental_dependency for parent_config in self.parent_stream_configs])
+        if not parent_state and not incremental_dependency:
             return
 
+        if not parent_state and incremental_dependency:
+            # Attempt to retrieve child state
+            child_state = list(stream_state.values())
```
Comment (bazarnov) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py:255]: Doesn't values() method provides an iter by default? is there a reason to wrap it within a list(), still?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py [hunk 2]
```diff
@@ -221,14 +223,46 @@ def set_initial_state(self, stream_state: StreamState) -> None:
...
+            child_state = child_state[0] if child_state else {}
+            parent_state = {}
+
+            # Copy child state to parent streams with incremental dependencies
+            if child_state:
```
Comment (bazarnov) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py:260]: nit: substream_state? (suggestion)

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py (hunk 1 above)
Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py:255]: It returns `dict_values`, and it is not subscriptable.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py (hunk 2 above)
Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py]: Fixed

**Review by bazarnov — APPROVED** [review]

**Comment by tolik0** [comment]
Zendesk-Support acceptance tests are failing due to the old CDK version, but regression tests are returning the same number of records. Here is the PR that updates the version: [PR Link](https://github.com/airbytehq/airbyte/pull/48379/files#diff-65515abe209173f187e7314081aea6ea6d852a272b18ee7e94feadedd4b6a23eR46)

**Comment by maxi297** [comment]
If it can help you test with zendesk-support, we have a PR ready for making in compatible with the latest CDK and concurrent: https://github.com/airbytehq/airbyte/pull/48379

**Review by maxi297 — APPROVED** [review]
LGTM! Thanks @tolik0

**Pull request merged**